### PR TITLE
Make cwrap Python3 compatible

### PR DESCRIPTION
**Pull request by JensGM**
In python 3, file streams are buffered on multiple levels which means
that it is no longer safe to extract a FILE* pointer from a python file
object. The old CFILE implementation can thus no longer be used. In
order to support function bindings that rely on the old python file to
FILE* mechanism, a thin compatibility layer is introduced.

char* strings from C can no longer be directly used as python str
objects. We therefor need to encode and decode str objects when passed
as arguments to or returned from foreign function calls.

The MethodType constructor in python 3 does not take a class as
argument. This is solved by deferring the construction to the functions
__get__ invocation, where we have access to the calling instance.

### CODE REVIEW

File: cwrap/cfile.py
Comment (joakim-hove) [inline cwrap/cfile.py]: Debug message `print(dir(self))`?

File: cwrap/cfile.py
Comment (joakim-hove) [inline cwrap/cfile.py]: Debug message

File: cwrap/cfile.py
Comment (joakim-hove) [inline cwrap/cfile.py]: The `fclose()` will flush the user buffers to the OS - no need to explicitly call `fflush()`

File: cwrap/cfile.py
Comment (joakim-hove) [inline cwrap/cfile.py]: Do you need to expose the `closed` property - I would hope not!

File: cwrap/cfile.py
Comment (joakim-hove) [inline cwrap/cfile.py]: `fclose()` should *not* fail - if it actually does I would say that exception material.

File: cwrap/prototype.py
Comment (joakim-hove) [inline cwrap/prototype.py]: Commented out code!

File: cwrap/prototype.py
Comment (joakim-hove) [inline cwrap/prototype.py]: Commented code!

File: cwrap/cfile.py
Comment (joakim-hove) [inline cwrap/cfile.py]: Debug print

File: cwrap/cfile.py
Comment (joakim-hove) [inline cwrap/cfile.py]: It should - IMO - work without the `'.6'` argument.

**Review by joakim-hove — DISMISSED** [review]

**Comment by joakim-hove** [comment]
I realize this is still very much WIP - but it looks good to me; and I am really looking forward to get it in.

Question: The new `open` implemention should be used as (for instance):
```
from cwrap.clib import open as copen

f = copen("file", "w")
....
```
I think I would prefer that it be named slightly differently - so that confusion with the built in Python `open()` is less probable?!

File: cwrap/prototype.py
Comment (JensGM) [reply cwrap/prototype.py]: This has changed a lot now, and will still. I'll take a look at these comments again when it's less WIP

File: cwrap/cfile.py
Comment (pgdr) [inline cwrap/cfile.py:145]: so `_fclose` returns an `int` and we return `not _fclose()` unless we are closed, in which case we return `None`?

Comment (JensGM) [reply cwrap/cfile.py:145]: The return value `0` from `fclose` means ok, by returning not of the result, we return a boolean that is true if we succeeded and false if not.

File: cwrap/prototype.py
Comment (pgdr) [inline cwrap/prototype.py]: Either remove or support, but I would hope we can deal with `char**` in Python3 as well?

Comment (JensGM) [reply cwrap/prototype.py]: I don't think we can support it very easily, but it seem to never be used.

Comment (pgdr) [reply cwrap/prototype.py]: Okay, so just delete the line.  Consider also the effect of deleting the corresponding line in PY2.

File: cwrap/cfile.py
Comment (pgdr) [reply cwrap/cfile.py:145]: And the None means it's already closed and thus had no effect?

File: cwrap/prototype.py
Comment (JensGM) [reply cwrap/prototype.py]: I've removed it

File: cwrap/cfile.py
Comment (JensGM) [reply cwrap/cfile.py:145]: Oh yea, I guess that it should return True if already closed? I've updated it to do so.

File: cwrap/prototype.py
Comment (pgdr) [reply cwrap/prototype.py]: Okay, just leave the Py2 as it is.

**Comment by pgdr** [comment]
This is now getting very close to mergable!

I'll let Joakim chime in one last time before we merge it.

File: cwrap/cfile.py
Comment (JensGM) [reply cwrap/cfile.py]: I'll change it to throw now

**Review by pgdr — APPROVED** [review]
Thank you, Jens, this is a very good contribution!

Hold off merge until Joakim greens it as well.

File: cwrap/cfile.py
Comment (joakim-hove) [inline cwrap/cfile.py:146]: My suggestion would be to completely remove the `_closed` attribute and instead just use the `c_ptr`:

```python
def close(self):
    if self.c_ptr:
         cs = self._fclose()
         if (cs != 0):
            raise IOError( )
         self.c_ptr = None
```

I am also very certain the `fflush()` call is not needed; that is handled by `fclose()`

**Review by joakim-hove — APPROVED** [review]
I think this is very good - and look forward to getting it in. I have one question about the `close( )` method - address that if you agree with me; otherwise this is good to go.

**Comment by joakim-hove** [comment]
Should we - as part of this bump the version number, 1.4 or maybe 2.0?

**Comment by pgdr** [comment]
> Should we - as part of this bump the version number, 1.4 or maybe 2.0?

This should be 1.4, yes, as it changes the API quite significantly, but remains backwards compatible.  Hats off to Jens for backwards compatibility here!

We should push a new release to pip, though, I cannot remember how we did it, nor who...

File: cwrap/cfile.py
Comment (JensGM) [reply cwrap/cfile.py:146]: Ok, I'll make those changes, wait for travis and merge

Comment (JensGM) [reply cwrap/cfile.py:146]: Nvm, using closed is much easier than checking if the file is valid. Calling close on a failed file causes a segfault.

Comment (joakim-hove) [reply cwrap/cfile.py:146]: OK - that is fine!

**Comment by joakim-hove** [comment]
> We should push a new release to pip, though, I cannot remember how we did it, nor who...

It seems the necessary swearing is already in the `.travis.yml`file?

**Comment by pgdr** [comment]
Wow, that's quite amazing.  There's even a 1.4 available on pypi already!?

**Comment by JensGM** [comment]
There is, but it calls it self 1.2.1 but is infact 1.4?? Theres some version number update missing I guess

**Pull request merged**
